File: rust/src/actions/entity.rs
use serde_json::{Value, json};

// use crate::hash::HashMap;
use crate::string::str;
use crate::context::Context;
use crate::options::Options;
// use crate::storage::Storage;
use crate::symbols::extract_symbol;
use crate::types::{
    Actions as ActionsObject,
    // EntityPayload,
};

/**
 * Struct representing actions that can be performed on entities.
 * This struct encapsulates the logic for creating, merging, and removing entities.
 */
pub trait EntityActions {//}

// impl EntityActions {
    /**
     * Creates a new entity in the current context.
     */
    fn create_entity(id: &Value, context: &Context, options: &Options) {
        let Options { compress_strings_as_ints, .. } = options;

        if id.is_null() {
            return;
        }

        if *compress_strings_as_ints {
            // TODO: Implement extract_symbol function
            let id = extract_symbol::<String>(id.as_i64().unwrap() as u32, context, options);
            if id.is_empty() {
                return;
            }
            // let id = Value::String(*id);
            context.create_entity(id, options);
        } else {
            context.create_entity(&id.as_str().unwrap().to_string(), options);
        }
    }

    /**
     * Retrieves entities from the current context.
     */
    fn entities(_: &Value, context: &Context, options: &Options) {
        let Options { responder, enum_default_symbols, .. } = options;

        // TODO: allow a queryies by compnent values using defined indexes, for example lookup actors by position

        // TODO: paginate large sets of actors

        let symbols = context.symbols;
        let merge_entities_symbol = symbols.find(str("mergeEntities")).unwrap();

        let entities = &context.get_entities(None, None);
        (responder)(&json!([merge_entities_symbol, entities]), None);
    }

    /**
     * Merges entities into the current context.
     */
    fn merge_entities(payload: &Value, context: &Context, options: &Options) {
        let array = payload.as_array().unwrap();
        let vec = array.iter().map(|v| str(v.as_str().unwrap())).collect::<Vec<&String>>();
        context.merge_entities(&vec, options);
    }

    /**
     * Removes an entity from the current context.
     */
    fn remove_entity(id: &Value, context: &Context, options: &Options) {
        let Options { compress_strings_as_ints, .. } = options;

        if id.is_null() {
            return;
        }

        if *compress_strings_as_ints {
            let id = extract_symbol::<String>(id.as_i64().unwrap() as u32, context, options);
            if id.is_empty() {
                return;
            }
            context.remove_entity(id, options);
        } else {
            context.remove_entity(&id.as_str().unwrap().to_string(), options);
        }
    }
}

struct Actions;

impl EntityActions for Actions {}

fn setup_actions<'a>(actions: &mut ActionsObject) -> &'a ActionsObject<'a> {
    actions.insert(
        &String::from("createEntity"),
        Box::new(&(Actions::create_entity as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("entities"),
        Box::new(&(Actions::entities as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("mergeEntities"),
        Box::new(&(Actions::merge_entities as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("removeEntity"),
        Box::new(&(Actions::remove_entity as fn(&Value, &Context, &Options))),
    );
    &actions
}

/**
 * An object that maps the names of actions to their corresponding methods in the EntityActions struct.
 */
pub const ACTIONS: &ActionsObject = setup_actions(&mut ActionsObject::new());

File: ecmascript/src/actions/entity.ts
import { Context } from '../context'
import { Emitter } from '../emitter'
import { Options } from '../options'
import { extractSymbol } from '../symbols'

/**
 * Creates a new instance of the EntityActions class that extends the provided Parent class
 * 
 * @param {any} Parent - The class to be extended by the EntityActions class. 
 * @returns {any} - A new class that extends the provided Parent class and the EntityActions class.
 */
export const EntityActionsFactory = (Parent: any = Object): any => class EntityActions extends Parent {
    /**
     * Creates a new entity in the current context.
     *
     * @param {any} id - The identifier for the entity to be created.
     * @param {Context} context - The current context in which the entity is to be created.
     * @param {Options | any} options - The options for creating the entity. If an instance of Options is not provided, a new one will be created.
     */
    createEntity(id: any, context: Context, options: Options | any): Promise<boolean> | boolean {  
        options = options = Options.ensure(options, this)
        const { compressStringsAsInts } = options

        if (id === undefined || id === null || id === '') { return false; }

        if (compressStringsAsInts) {
            id = extractSymbol(id, context, options)
            if (id === '') { return false; }
        }

        return context.createEntity(id, options)
    }

    /**
     * Retrieves entities from the current context.
     *
     * @param {any} payload - This parameter is not used in the method.
     * @param {Context} context - The current context from which the entities are retrieved.
     * @param {OptionsExtended | any} options - The options for retrieving entities. If an instance of Options is not provided, a new one will be created.
     */
    entities(payload: any, context: Context, options: Options | any) {
        options = options = Options.ensure(options, this)
        const { responder, isAuthority, isAsyncStorage, pageSize, enableQuerying, enumDefaultSymbols } = options

        if (!isAuthority) { return }

        const sendEntities = (pages: string[][]) => {
            for (const page of pages) {
                responder([enumDefaultSymbols.mergeEntities, page])
            }
        }

        const ctxEntities = context.getEntities(enableQuerying ? payload : null, pageSize)

        if (isAsyncStorage) {
            (ctxEntities as Emitter<string[][]>).emitTo(sendEntities, true)
        } else {
            sendEntities(ctxEntities as string[][])
        }
    }

    /**
     * Merges entities into the current context.
     *
     * @param {any[]} payload - The payload containing the entities to be merged.
     * @param {Context} context - The current context in which the entities are to be merged.
     * @param {OptionsExtended | any} options - The options for merging. If an instance of Options is not provided, a new one will be created.
     */
    mergeEntities(payload: any[], context: Context, options: Options | any) {
        options = options = Options.ensure(options, this)

        context.mergeEntities(payload, options)
    }

    /**
     * Removes an entity from the current context.
     *
     * @param {any} id - The identifier for the entity to be removed.
     * @param {Context} context - The current context from which the entity is to be removed.
     * @param {OptionsExtended | any} options - The options for removing the entity. If an instance of Options is not provided, a new one will be created.
     */
    removeEntity(id: any, context: Context, options: Options | any) {
        options = options = Options.ensure(options, this)
        const { compressStringsAsInts } = options

        if (id === undefined || id === null || id === '') { return }

        if (compressStringsAsInts) {
            id = extractSymbol(id, context, options)
            if (id === '') { return }
        }

        context.removeEntity(id, options)
    }
}

/**
 * Class representing actions that can be performed on entities.
 * This class encapsulates the logic for creating, merging, and removing entities.
 */
export class EntityActions extends EntityActionsFactory() { }

const __EntityActions__ = new EntityActions()

/**
 * An object that maps the names of actions to their corresponding methods in the EntityActions class.
 */
export const actions = {
    /**
     * Creates a new entity in the current context.
     */
    createEntity: __EntityActions__.createEntity,

    /**
     * Retrieves entities from the current context.
     */
    entities: __EntityActions__.entities,

    /**
     * Merges entities into the current context.
     */
    mergeEntities: __EntityActions__.mergeEntities,

    /**
     * Removes an entity from the current context.
     */
    removeEntity: __EntityActions__.removeEntity
}

export default actions